Feat - Voting Power

akyo8 opened the pull request.

Commit:
Voting Power Calc - Uniswap, SushiSwap, Balancer, Wallet

Commit:
borrowed, staked balances in pools

Commit:
Add test script, update interfaces, hardhat configuration

Commit:
update lending supplied, borrowed methods

jeremyHD commented on contracts/CreamVotingPower.sol:
```diff
@@ -1,12 +1,182 @@
+    address public constant balancer =
+        0x280267901C175565C64ACBD9A3c8F60705A72639;
+
+    uint256 public MINIMUM_VOTING_POWER = 1;
```
should be 1e18

jeremyHD commented on contracts/CreamVotingPower.sol:
```diff
@@ -1,12 +1,182 @@
+
+    uint256 public MINIMUM_VOTING_POWER = 1;
+
+    function getVotingPower(address _holder) public view returns (uint256) {
```
We'll like to conform the standard ERC20 interface for this function signature
`balanceOf`

jeremyHD commented on scripts/sample-script.js:
```diff
@@ -1,32 +1,43 @@
+  for (let i = 0; i < addresses.length; i++) {
+    try {
+      const response = await VotingPower.getVotingPower(addresses[i]);
+      console.log(`=========${addresses[i]}=========\n${response}`);
```
Oh nice, you have written the test that I was trying to tell you about! Sorry I didn't see this just now

jeremyHD commented on scripts/example.json:
```diff
@@ -0,0 +1,9 @@
+  "0x56e0150b3cbb9f191fe360740fcfa2edbd1b3697",
+  "0xbbe21151c726F9EEc42516524f58D0057261CC8a",
+  "0x6bD48938B9F9Dd02ee3D6dccd0379DDF0af22D38",
+  "0xa805cb6907130c0c7F3534Ee2EBc62D806FFB820"
```
And you also bring the same addresses from the examples. Good Job!!!

If you set the `periods` in [example.json](https://github.com/snapshot-labs/snapshot.js/blob/master/src/strategies/cream/examples.json#L15) to `0` for snapshot strategy.
And run `npm run test --strategy=cream` on snapshot.js repo, the result should be the same or close enough

akyo8 replied:
@jeremyHD 
When `periods` was set `0`, it ran into this issue `Error: division by zero`.
Please tell me if you have any idea with it. Thanks

jeremyHD replied:
set to 1. My bad

jeremyHD commented on contracts/CreamVotingPower.sol:
```diff
@@ -1,12 +1,182 @@
+            "VotingPower.lendingSupply: Zero Address"
+        );
+
+        uint256 totalLending = ICToken(crCream).balanceOf(_holder) * ICToken(crCream).exchangeRateStored() / 1e18;
```
```suggestion
        uint256 totalLending = ICToken(crCream).balanceOf(_holder) * ICToken(crCream).exchangeRateStored().div( 1e18);
```

jeremyHD commented on contracts/CreamVotingPower.sol:
```diff
@@ -1,12 +1,182 @@
+        require(
+            votingPower > MINIMUM_VOTING_POWER,
+            "VotingPower.getVotingPower: Invalid"
+        );
```
return 0 if votingPower < Min voting power

Commit:
defiPulse amount adding, minor issues

akyo8 commented:
@jeremyHD 
I updated with fixing of above comments and added defiPulse amount.
If you have any improvements or suggestions, please let me know. thanks

jeremyHD commented on contracts/CreamVotingPower.sol:
```diff
@@ -1,12 +1,202 @@
+        uint256 rewardDebt;
+    }
+
+    address public constant defiPulse = 0x1494CA1F11D487c2bBe4543E90080AeBa4BA3C2b;
```
should be defiPulseIndex or just name it dpi

akyo8 replied:
it's been named defiPulseIndex

jeremyHD commented on contracts/CreamVotingPower.sol:
```diff
@@ -1,12 +1,202 @@
+
+        uint256 totalLending = ICToken(crCream).balanceOf(_holder).mul(ICToken(crCream).exchangeRateStored()).div(1e18);
+        uint256 borrowed = ICToken(crCream).borrowBalanceStored(_holder);
+
```
Please add a sanity check here
```
if (borrowed > totalLending) {
    return 0
}
```

or else safeMath will revert if underflow

akyo8 replied:
added sanity check for it

jeremyHD commented on contracts/CreamVotingPower.sol:
```diff
@@ -1,12 +1,202 @@
+        uint256 totalStaked = 0;
+
+        for (uint256 i = 0; i < 4; i++) {
+            totalStaked.add(ILPool(lPool[i]).balanceOf(_holder));
```
```suggestion
            totalStaked = totalStaked.add(ILPool(lPool[i]).balanceOf(_holder));
```

akyo8 replied:
fixed

Commit:
voting power improvement - sanity-check, staked balance sum issue, defiPulse naming

akyo8 commented:
@jeremyHD 
Please check above fixing and this is final results of testing

=========0xeB8A4eADd643Ea3873dfecFBa18BD709D3206919=========
25988067596870823513528
=========0xe0Ba6c4c82376c76386D1Ca2ae2516Ab18882e3C=========
443299961957519273343
=========0xE32F576578F9aea942226B6A280fbCe93da45638=========
0
=========0x56e0150b3cbb9f191fe360740fcfa2edbd1b3697=========
83128575877236447792
=========0xbbe21151c726F9EEc42516524f58D0057261CC8a=========
0
=========0x6bD48938B9F9Dd02ee3D6dccd0379DDF0af22D38=========
51222026950618754185
=========0xa805cb6907130c0c7F3534Ee2EBc62D806FFB820=========
0
